src/main.c

```diff
@@ -351,18 +362,120 @@ key_binding_func_key (MBWindowManager   *wm,
   hd_dbus_send_event (s);
 }
 
+/* Close all input devices previously opened */
+static void
+close_input_devices (Display *dpy)
+{
+  int i;
+
+  for (i = 0; i < xi_devices->len; i++)
+    {
+      hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, i);
+
+      if (xi_dev->dev)
+        {
+          XCloseDevice (dpy, xi_dev->dev);
+          xi_dev->dev = NULL;
+        }
+    }
+}
+
+/* Fill in the array with input devices type is touchscreen */
+static void
+enumerate_input_devices (Display *dpy)
+{
+  static Atom atom_touchscreen = None;
+  XDeviceInfo *devinfo;
+  int i, ndev;
+  GArray *eclass;
+
+  if (xi_devices)
+    {
+      close_input_devices (dpy);
+      g_array_free (xi_devices, TRUE);
+      xi_devices = NULL;
+    }
+
+  if (atom_touchscreen == None)
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen == None)
+    return;
+
+  /* get XInput DeviceMotion events */
+  devinfo = XListInputDevices (dpy, &ndev);
+
+  eclass = g_array_new (FALSE, FALSE, sizeof (XEventClass));
+  xi_devices = g_array_new (FALSE, TRUE, sizeof (hd_xi_device));
+
+  for (i = 0; i < ndev; i++)
+    {
+      XDeviceInfo info = devinfo[i];
+
+      if (info.use == IsXExtensionPointer)
+        {
+          XEventClass ev_class;
+          XDevice *dev = XOpenDevice (dpy, info.id);
+          XID id = info.id;
+
+          if (xi_devices->len <= id)
+            g_array_set_size(xi_devices, id + 1);
+
+          hd_xi_device *xi_dev = &g_array_index (xi_devices, hd_xi_device, id);
+
+          DeviceMotionNotify (dev, xi_motion_ev_type, ev_class);
+          g_array_append_val (eclass, ev_class);
+
+          xi_dev->is_ts = (info.type == atom_touchscreen);
+          xi_dev->dev = dev;
+#if 0
+          g_warning ("### %d/%u. %s(%d):%s.%d t:%d c:%d\n", i, eclass->len,
```

Comment:
Don't care much, but did you want to keep this here, in '#if 0' ?

Reply:
It is very useful for debugging